### PR TITLE
AC_Fence: add a pre-arm check for polygon fence validity

### PR DESCRIPTION
**Pull request by peterbarker**
Also change SYS_STATUS bit to be unhealthy if boundary invalid

### CODE REVIEW

File: libraries/AC_Fence/AC_Fence.cpp [hunk 1]
```diff
@@ -545,5 +566,11 @@ bool AC_Fence::geofence_failed() const
     if (get_breaches() != 0) {
         return true;
     }
+    if (_enabled_fences & AC_FENCE_TYPE_POLYGON) {
+        if (!_boundary_valid) {
+            return true;
```
Comment (OXINARF) [inline libraries/AC_Fence/AC_Fence.cpp]: I don't like this as it is contrary to what the breach code does: if the boundary isn't valid, then it says that not breach occurred - so why would we report a failed geofence?

By the way, the term geofence here was badly applied, we call it polygon fence everywhere except these new methods.

**Comment by OXINARF** [comment]
This closes #7138.

**Comment by peterbarker** [comment]
On Thu, 14 Dec 2017, Francisco Ferreira wrote:

> 
> In libraries/AC_Fence/AC_Fence.cpp:
> 
> > @@ -545,5 +566,11 @@ bool AC_Fence::geofence_failed() const
>      if (get_breaches() != 0) {
>          return true;
>      }
> +    if (_enabled_fences & AC_FENCE_TYPE_POLYGON) {
> +        if (!_boundary_valid) {
> +            return true;
> 
> I don't like this as it is contrary to what the breach code does: if the
> boundary isn't valid, then it says that not breach occurred - so why would
> we report a failed geofence?

You can't breach a fence which is malformed.  One problem that would arise 
if we were to say that if the fence is malformed you are always breached 
is that updating the fence in-flight would probably always lead to a 
breach - we'd need some weird logic to sort that out.

If you are expecting a fence to work (by having it in _enabled_fences) and 
it malformed then it is definitely unhealthy and should be marked in 
sys_status as such).

That was my logic, anyway!

Frankly, we need to go back and overhaul this stuff at some stage.

> By the way, the term geofence here was badly applied, we call it polygon
> fence everywhere except these new methods.

These sys_staus helpers cover all of the fence types, not just polyfence? 
I chose "geofence_" to be consistent with the same methods in Plane.

**Comment by OXINARF** [comment]
> These sys_staus helpers cover all of the fence types, not just polyfence?
I chose "geofence_" to be consistent with the same methods in Plane.

Ah my bad! Omitting it would have been nicer then, `fence.failed()` looks better than `fence.geofence_failed()`.

> If you are expecting a fence to work (by having it in _enabled_fences) and
it malformed then it is definitely unhealthy and should be marked in
sys_status as such).
>
> That was my logic, anyway!

OK, I understand your logic, but then there are other cases where fence will fail because it can't be applied (like all the pre-arm check cases).

**Comment by amilcarlucas** [comment]
I also vote to omit the word geofence here.

**Comment by peterbarker** [comment]
@amilcarlucas @OXINARF Omitting the word "geofence" leads to a name conflict; we use "enabled" for something else entirely in AC_Fence.

We could, perhaps, use `sys_status_enabled()`, `sys_status_present()` and `sys_status_failed()`?

**Comment by OXINARF** [comment]
> We could, perhaps, use sys_status_enabled(), sys_status_present() and sys_status_failed()?

@peterbarker Sounds good. Just to be clear, that's not what is holding me up on this PR - it's the sys status failed report only covering the polygon; in my opinion, we either follow what the breach code does and don't care or, if we want to report a failure when a breach can't be verified, then we should report it in all cases.

**Comment by peterbarker** [comment]
On Tue, 26 Dec 2017, Francisco Ferreira wrote:

>       We could, perhaps, use sys_status_enabled(),
>       sys_status_present() and sys_status_failed()?
> 
> @peterbarker Sounds good. Just to be clear, that's not what is holding me
> up on this PR - it's the sys status failed report only covering the
> polygon; in my opinion, we either follow what the breach code does and

Are you suggesting we should also check that if the circle fence is 
enabled it has a non-zero radius?  Or that altitude is positive in the 
case of the height fence?

**Comment by OXINARF** [comment]
> Are you suggesting we should also check that if the circle fence is
enabled it has a non-zero radius?  Or that altitude is positive in the
case of the height fence?

Those are great ones (we should probably change them to pre-arm checks instead of setting their value on construction, which is wrong given the saved values haven't been loaded), but I was thinking more of checking we have position for the circle one, for example. It will continue the check for breach with the old *distance to home*, but that distance may be completely wrong, so why wouldn't the report say it is failed too?

**Comment by peterbarker** [comment]
@OXINARF I've pushed up more bits-and-pieces for the health bit.  Fence sensor goes unhealthy if we don't have position or the parameters are bad.

**Comment by OXINARF** [comment]
Merged, thanks!

**Pull request merged**